### PR TITLE
[persist] Spill long runs to separate blobs in S3

### PR DESCRIPTION
**Pull request by bkirwi**
See comments in https://github.com/MaterializeInc/database-issues/issues/7975 for important background and design.

The core change in this PR is
[Add RunPart enum and HollowRuns](https://github.com/MaterializeInc/materialize/pull/30094/commits/e0214a95d4d500f10b2fd72e12baad9b897a860e), which:
- Adds a new `HollowRun` type with a list of parts, and a `HollowRunRef` type with a pointer and a small amount of denormalized metadata.
- Adds a `RunPart` enum with variants for that and the existing `BatchPart` enum.
- Adapts ~all of persist to handle the new variant, including adapting many iterators to streams and otherways async-ifying things.

### Motivation

Implements: https://github.com/MaterializeInc/database-issues/issues/8401

### Tips for reviewer

It's a big PR! I attempted to split up a few chunkier changes into their own commits, but there's still a 500-liner that adds the new enum/variant and adapts many of the smaller usages. Let me know if it's worth taking time to walk through anything live.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

**Comment by danhhz** [comment]
:fire: I'll be a little slammed through the weekly demo tomorrow (I'm giving it), but I'll take a look after that!

**Comment by bkirwi** [comment]
No rush! Nontrivial and not urgent.

**Comment by bkirwi** [comment]
There's what looks like a stack overflow in `test_no_block`... I think I've seen this before when a future started taking up more stack space? Will look into it.

**Comment by bkirwi** [comment]
Nightlies: https://buildkite.com/materialize/nightly/builds/10124

Clean except for some upgrade failures that have been fixed in main since I branched off. I'll rerun once I've got the fix for the unit test up.

**Comment by bkirwi** [comment]
Clean nightly build: https://buildkite.com/materialize/nightly/builds/10147

File: src/persist-client/src/iter.rs [hunk 1]
```diff
@@ -621,8 +632,18 @@ where
         let mut ready_futures: FuturesUnordered<_> = self.runs[0..first_larger]
             .iter_mut()
             .map(|run| async {
-                let part = &mut run.front_mut().expect("trimmed run should be nonempty").0;
-                if let ConsolidationPart::Queued { data, task } = part {
+                // It's possible for there to be multiple layers of indirection between us and the first available encoded part:
+                // if the first part is a `HollowRuns`, we'll need to fetch both that and the first part in the run to have data
+                // to consolidate. So: we loop, and bail out of the loop when either the first part in the run is available or we
+                // hit some unrecoverable error.
+                'fetch_loop: loop {
```
Comment (danhhz) [inline src/persist-client/src/iter.rs:639]: The control flow here is a bit hard to follow (also I usually find naming loops to be a code smell). Is it possible to pull this out into a method? (Fine if the answer is "no" or "not worth it")

File: src/persist-client/src/batch.rs [hunk 2]
```diff
@@ -1338,48 +1338,84 @@ pub(crate) fn validate_truncate_batch<T: Timestamp>(
...
-        let () = deletes.collect().await;
     }
-}
 
-impl Deref for PartDeletes {
-    type Target = BTreeSet<PartialBatchKey>;
-    fn deref(&self) -> &Self::Target {
-        &self.0
+    pub fn is_empty(&self) -> bool {
+        self.hollow_runs.is_empty() && self.hollow_parts.is_empty()
```
Comment (danhhz) [inline src/persist-client/src/batch.rs:1498]: nit: Maybe deconstruct `Self{hollow_runs, hollow_parts} = self` to be defensive against another field being added later?

File: src/persist-client/src/batch.rs [hunk 3]
```diff
@@ -1338,48 +1338,84 @@ pub(crate) fn validate_truncate_batch<T: Timestamp>(
...
+            let Some((run_key, run_ref)) = self.hollow_runs.pop_first() else {
+                break;
+            };
+
+            if let Some(run) = run_ref.get(shard_id, blob).await {
+                // Queue up both all the individual parts and the run itself for deletion.
+                for part in &run.parts {
+                    self.add(part);
+                }
+                self.hollow_parts.insert(run_key);
```
Comment (danhhz) [inline src/persist-client/src/batch.rs:1536]: Should we rename `hollow_parts`?

File: src/persist-client/src/batch.rs [hunk 4]
```diff
@@ -486,6 +494,7 @@ impl BatchBuilderConfig {
             record_run_meta,
             expected_order,
             structured_key_lower_len: STRUCTURED_KEY_LOWER_LEN.get(value),
+            run_length_limit: MAX_RUN_LEN.get(value).clamp(2, usize::MAX),
```
Comment (danhhz) [inline src/persist-client/src/batch.rs:498]: TIL `clamp`, cool

File: src/persist-client/src/batch.rs [hunk 5]
```diff
@@ -977,8 +1018,13 @@ pub(crate) struct BatchParts<T> {
     lower: Antichain<T>,
     blob: Arc<dyn Blob>,
     isolated_runtime: Arc<IsolatedRuntime>,
-    writing_parts: VecDeque<JoinHandle<RunPart<T>>>,
-    finished_parts: Vec<RunPart<T>>,
+    next_index: u64,
+    /// A hierarchy of runs. `writing_runs[0]` is a run of individual batch parts;
```
Comment (danhhz) [inline src/persist-client/src/batch.rs:1023]: :100:

**Review by danhhz — APPROVED** [review]
LGTM! I quite appreciate the commit structure, make it much easier to review

Left a few comments, but the only ones that I'd say are blocking are the exhaustive match ones

File: src/persist-client/src/cli/bench.rs [hunk 6]
```diff
@@ -76,8 +76,8 @@ async fn bench_s3(args: &S3FetchArgs) -> Result<(), anyhow::Error> {
         let mut fetches = Vec::new();
         for part in snap.iter().flat_map(|x| x.parts.iter()) {
             let key = match part {
-                BatchPart::Hollow(x) => x.key.complete(&shard_id),
-                BatchPart::Inline { .. } => continue,
+                RunPart::Single(BatchPart::Hollow(x)) => x.key.complete(&shard_id),
+                _ => continue,
```
[the remark this answers is not in the document]
Comment (bkirwi) [reply src/persist-client/src/cli/bench.rs:80]: Sure! I will replace this with just a fetch for the hollow run for now, since doing a recursive fetch seems to go against the spirit of this benchmark... but let me know if you prefer otherwise.

File: src/persist-client/src/internal/state_versions.rs [hunk 7]
```diff
@@ -364,7 +364,7 @@ impl StateVersions {
                         }
                         // TODO: Would be nice to include these too, but we lose the
                         // info atm.
-                        BatchPart::Inline { .. } => None,
+                        _ => None,
```
[the remark this answers is not in the document]
Comment (bkirwi) [reply src/persist-client/src/internal/state_versions.rs:367]: Good callout - updating!

File: src/persist-client/src/cli/bench.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (danhhz) [reply src/persist-client/src/cli/bench.rs:80]: It's meant to be recursive! Totally fine to move the flattening one top in the setup, though, and just have a list of blob keys to fetch here

[the remark this answers is not in the document]
Comment (bkirwi) [reply src/persist-client/src/cli/bench.rs:89]: Aight, done!

File: src/persist-client/src/iter.rs (hunk 1 above)
Comment (bkirwi) [reply src/persist-client/src/iter.rs:639]: I did try this and got stuck, so I'm hesitant to try again. 😅

I can swap those labelled breaks for returns, though, if you prefer it!

File: src/persist-client/src/batch.rs (hunk 4 above)
Comment (bkirwi) [reply src/persist-client/src/batch.rs:498]: Yeah... basically just `value.max(lower_bound).min(upper_bound)`, but I do find it easier to follow for whatever reason.

File: src/persist-client/src/iter.rs (hunk 1 above)
Comment (bkirwi) [reply src/persist-client/src/iter.rs]: (I did the swap, since it seems at least as clear and a bit more normal...)
